SmartDeviceLink/private/SDLPresentAlertOperation.m

```diff
@@ -0,0 +1,463 @@
+
+@interface SDLAlertAudioData()
+
+@property (nullable, copy, nonatomic, readonly) NSDictionary<SDLFileName *, SDLFile *> *audioFileData;
+
+@end
+
+@interface SDLAlertView()
+
+/// Handler called when the alert should be dismissed.
+@property (copy, nonatomic) SDLAlertCanceledHandler canceledHandler;
+
+@end
+
+@interface SDLSoftButtonObject()
+
+/// Unique id assigned to the soft button.
+@property (assign, nonatomic) NSUInteger buttonId;
+
+@end
+
+@interface SDLPresentAlertOperation()
+
+@property (strong, nonatomic) NSUUID *operationId;
+@property (weak, nonatomic) id<SDLConnectionManagerType> connectionManager;
+@property (weak, nonatomic) SDLFileManager *fileManager;
+@property (weak, nonatomic) SDLSystemCapabilityManager *systemCapabilityManager;
+@property (strong, nonatomic, readwrite) SDLAlertView *alertView;
+@property (assign, nonatomic) UInt16 cancelId;
+@property (copy, nonatomic, nullable) NSError *internalError;
+
+@end
+
+@implementation SDLPresentAlertOperation
+
+#pragma mark - Lifecycle
+
+- (instancetype)initWithConnectionManager:(id<SDLConnectionManagerType>)connectionManager fileManager:(SDLFileManager *)fileManager systemCapabilityManager:(SDLSystemCapabilityManager *)systemCapabilityManager currentWindowCapability:(nullable SDLWindowCapability *)currentWindowCapability alertView:(SDLAlertView *)alertView cancelID:(UInt16)cancelID {
+
+    self = [super init];
+    if (!self) { return nil; }
+
+    _connectionManager = connectionManager;
+    _fileManager = fileManager;
+    _systemCapabilityManager = systemCapabilityManager;
+
+    __weak typeof(self) weakSelf = self;
+    alertView.canceledHandler = ^{
+        [weakSelf sdl_cancelInteraction];
+    };
+    _alertView = [alertView copy];
+
+    _cancelId = cancelID;
+    _operationId = [NSUUID UUID];
+    _currentWindowCapability = currentWindowCapability;
+
+    return self;
+}
+
+- (void)start {
+    [super start];
+    if (self.isCancelled) { return; }
+
+    NSError *alertViewValidatedError = [self sdl_isValidAlertViewData:self.alertView];
+    if (alertViewValidatedError) {
+        [self finishOperation];
+        self.internalError = alertViewValidatedError;
+        return;
+    }
+
+    dispatch_group_t uploadFilesTask = dispatch_group_create();
+    dispatch_group_enter(uploadFilesTask);
+
+    dispatch_group_enter(uploadFilesTask);
+    [self sdl_uploadImagesWithCompletionHandler:^{
+        dispatch_group_leave(uploadFilesTask);
+    }];
+
+    dispatch_group_enter(uploadFilesTask);
+    [self sdl_uploadAudioFilesWithCompletionHandler:^{
+        dispatch_group_leave(uploadFilesTask);
+    }];
+
+    dispatch_group_leave(uploadFilesTask);
+     
+    // This will always run after all `leave`s
+    __weak typeof(self) weakSelf = self;
+    dispatch_group_notify(uploadFilesTask, [SDLGlobals sharedGlobals].sdlConcurrentQueue, ^{
+        __strong typeof(weakSelf) strongSelf = weakSelf;
+        [strongSelf sdl_presentAlert];
```

Comment:
I thought about that but it seems redundant since either the alert does not have images/audio files and the cancel check was already performed or the cancel occurred during the image/file upload, which is already handled in those methods.